Add SPI support for Atmega based chips.

schmidtw opened the pull request:
This is based on @Nerzal's #1398 PR, but is a bit of a rework and
expansion to all the Atmega based chips.

schmidtw commented:
This is my first bit of coding with this code base & I'm not entirely sure I have the code in the right places.  I am fairly certain that the SPI port works (I have it connected to my [Saleae](https://www.saleae.com) and am able to confirm that the program below produces the right output using an Arduino Uno board.

```go
package main

import (
	"fmt"
	"machine"
	"time"
)

func main() {
	time.Sleep(time.Millisecond * 1000)
	fmt.Printf("hello world!\n")

	time.Sleep(time.Millisecond * 1000)

	cs := machine.SPI0_CS_PIN
	cs.High()
	cs.Configure(machine.PinConfig{Mode: machine.PinOutput})

	machine.SPI0.Configure(machine.SPIConfig{
		LSBFirst:  true,
		Frequency: 5,
		Mode:      0, //machine.Mode0,
	})
	out := []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 0xf0, 0x0f, 0xff}

	for {
		time.Sleep(time.Millisecond * 1000)

		fmt.Printf("SPI Out!\n")
		cs.Low()
		machine.SPI0.Tx(out, nil)
		cs.High()
	}
}
```

schmidtw commented:
I'm not sure where the right place to declare `SPI0` at.  If I do not declare it [here](https://github.com/tinygo-org/tinygo/pull/1614/files#diff-d1606be86f97c2b5674e97efa76e43e1b2a6cb95923b64b279e40050865dce86R56) when I try to compile the example program using `tinygo flash -target arduino spi.go` the SPI0 port is not defined.

Suggestions?

Nerzal commented:
Just out of curiosity - can you tell me what was wrong with my code? I did not manage to get it to run correctly.

And just to be sure - have you made sure that your "tinygo flash ..." commands refers to the correct tinygo version, that you have built? 
What i do most of the time to make sure it uses the correct tinygo binary, i build tinygo with the -o flag  like

```bash
 go build -o my_tinygo
```

and then directly use that binary like in the following example: 

```bash
./my_tinygo flash --target=arduino path/to/main.go
```

I think that could be your issue, because the placement of the SPI0 looks correct to me :)

Nerzal commented on src/machine/machine_atmega.go:
```diff
@@ -155,3 +155,91 @@ func (uart UART) WriteByte(c byte) error {
+	if config.Frequency == 0 {
+		config.Frequency = 4000000
+	}
+	if 0 == config.SCK {
```
You might want to resolve the Yoda Conditions, by switching the order of the check. 

Example: config.SCK == 0  instead of 0 == config.SCK

Nerzal commented on src/machine/machine_atmega.go:
```diff
@@ -155,3 +155,91 @@ func (uart UART) WriteByte(c byte) error {
+	case frequencyDivider >= 8:
+		avr.SPCR.SetBits(avr.SPCR_SPR0)
+		avr.SPSR.SetBits(avr.SPSR_SPI2X)
+	case frequencyDivider >= 4:
```
golang has no default fall through switch cases. 
So i fear, that you either need to add the fallthrough statement in that case, or just omit this case, to let it go through the default one.

schmidtw replied:
Good call out - I'm going to add a comment in the code to make this a bit more clear.  The `>= 4` case results in the bits being cleared to 0 which means no setting is needed for this case.  The default value is anything faster than 4 & is configured differently.

Nerzal reviewed (changes requested):
Mostly looks good to me, only found 2 little things :+1:

schmidtw commented:
> Just out of curiosity - can you tell me what was wrong with my code? I did not manage to get it to run correctly.

Good question.  I had to go back to try to figure this out.  The one thing I know is an issue is the way you set the clock.  The bits for the speed were using the bitmask of the register, but the SPI_2XCLOCK_MASK is 0x01 of the SPSR register and SPI_CLOCK_MASK is 0x03 of the SPCR register.  This means that for SPI_CLOCK_FOSC128 (3), your code was setting SPR0, SPR1 & 2XCLOCK when only SPR0 & SPR1 should have been set.

```
avr.SPSR.SetBits(clockDivider & SPI_2XCLOCK_MASK)
...
avr.SPCR.SetBits(avr.SPCR_SPE | avr.SPCR_MSTR | clockDivider&SPI_CLOCK_MASK)
```

Beyond that I'm not sure if there was much that was actually wrong.  I couldn't get the code to work with my initial test program, saw the clock issue and then pursued the approach to explicitly zero out all the registers used to make sure they were initialized to a known value and work from there.  After longer than I would like to admit I discovered that in my test program I was using the LED pin on the Arduino to signal when I should be getting a small burst of SPI... and the LED is the same pin as the SCK.  Once I removed the LED code everything worked.  By this time the code was very close to what I submitted.  I tried to normalize the code to match the patterns of what the code around it and your patch were after.


> And just to be sure - have you made sure that your "tinygo flash ..." commands refers to the correct tinygo version, that you have built?

Yes, but good question.

> What i do most of the time to make sure it uses the correct tinygo binary, i build tinygo with the -o flag like
> 
> ```shell
>  go build -o my_tinygo
> ```
> 
> and then directly use that binary like in the following example:
> 
> ```shell
> ./my_tinygo flash --target=arduino path/to/main.go
> ```
> 
> I think that could be your issue, because the placement of the SPI0 looks correct to me :)

Yeah, the command line you suggest works for building the test program when I have SPI0 where it is in the PR.  However, that definition seems to conflict with what is [here](https://github.com/tinygo-org/tinygo/blob/cca0eab3da7065d568ca073df61f03beecbd2086/src/machine/machine_generic.go#L8), which breaks the CI build.

Nerzal commented:
@schmidtw thanks for pointing out, what was broken in my PR. 
I'll just close my PR in favor to this one :)

aykevl commented on src/machine/board_arduino.go:
```diff
@@ -43,3 +43,11 @@ const (
+	SPI0_SCK_PIN Pin = PB5
+	SPI0_SDO_PIN Pin = PB3
+	SPI0_SDI_PIN Pin = PB4
+	SPI0_CS_PIN  Pin = PB2
```
I see that some other boards also have a defined CS pin, but to be honest I don't think it should be used like this. The whole point of the chip select pin is to select which (of multiple) external devices are currently in use, and therefore you should use one per external device, not one per SPI peripheral.

```suggestion
```

schmidtw replied:
I'm fine removing the CS pin.

I initially went with the various documents from Atmel (all have a CS or SS pin defined even though they're just GPIO).  In the past AVR32 or PIC chips I've seen have had a special group of CS pins so that the SPI hardware could do the enable/disable automatically based on which chip you wanted to send/receive from ... all via interrupts.  From what I can tell, that pattern doesn't exist in any of the machines tinygo supports.

schmidtw replied:
I moved the implementation up to atmega since from what I could tell all the atmega chips have the same SPI hardware register implementation.

I agree with the pins being defined per chip or board or by the user.  It's just really nice to be able to wire & when there's only one possible choice it "just works".  This seems consistent with I2C and USART as well.

aykevl replied:
> In the past AVR32 or PIC chips I've seen have had a special group of CS pins so that the SPI hardware could do the enable/disable automatically based on which chip you wanted to send/receive from ... all via interrupts. From what I can tell, that pattern doesn't exist in any of the machines tinygo supports.

Yeah, I'm aware this exists. We might want to add support for that eventually, but would need to find a way to express it in a portable way. And in any case, it only makes sense once DMA is supported (which it currently isn't, for the most part).

> I moved the implementation up to atmega since from what I could tell all the atmega chips have the same SPI hardware register implementation.

No. The atmega1284p for example has 3 SPI peripherals, so it can't use the exact same registers: there need to be 3 sets of them with different names. Maybe it has the same names for SPI0 but it can't have the same names for SPI1 and SPI2.

schmidtw replied:
I must be missing something, but I don't see the other 2 SPI peripherals in the datasheet.  I see the USART as SPI (many of the chips have that), but I didn't try to address those other 2 modes.

schmidtw replied:
I found a different supported chip that has the pattern you mentioned. `atmega1284p`

My larger question is would you rather have a bunch of duplicated code for each processor or have a more generic solution that works for all the Atmega chips?

schmidtw replied:
I found a different chip that matches your example: `atmega384pb`

My question is would you rather have a more generic implementation that all the Atmega chips can use, or do you want the code duplicated for each chip?

aykevl replied:
I'm referring to this product page: https://www.microchip.com/wwwproducts/en/ATMEGA1284P

But maybe you're right, with a quick scan I can't find other SPI buses. Which kind of makes sense, as you can't really do much with multiple SPI peripherals if you don't also use DMA (to work with them in parallel).

aykevl replied:
> My question is would you rather have a more generic implementation that all the Atmega chips can use, or do you want the code duplicated for each chip?

Ideally a generic solution if it is possible, but maybe it is not possible. For example, when I was working on a refactor of PWM support I couldn't really come up with a working generic solution.

aykevl commented on src/machine/machine_atmega.go:
```diff
@@ -155,3 +155,94 @@ func (uart UART) WriteByte(c byte) error {
+	if config.Frequency == 0 {
+		config.Frequency = 4000000
+	}
+	if config.SCK == 0 {
+		config.SCK = SPI0_SCK_PIN
+	}
+	if config.SDO == 0 {
+		config.SDO = SPI0_SDO_PIN
+	}
+	if config.SDI == 0 {
+		config.SDI = SPI0_SDI_PIN
+	}
```
This doesn't make sense to me. AVR chips have assigned fixed pins to certain SPI peripherals, so there is no point in configuring different pins.

schmidtw replied:
I based this on the prior PR as well as trying to be consistent with the other SPI implementations for consistency.  I'm fine deleting code that isn't used.

aykevl commented on src/machine/machine_atmega.go:
```diff
@@ -155,3 +155,94 @@ func (uart UART) WriteByte(c byte) error {
+	avr.SPSR.Set(0)
+
+	// Setup pins output configuration
+	config.SCK.Configure(PinConfig{Mode: PinOutput})
+	config.SDI.Configure(PinConfig{Mode: PinInput})
+	config.SDO.Configure(PinConfig{Mode: PinOutput})
```
Same here.

schmidtw replied:
Something needs to configure the pins for output mode.  If there is only one & the user asks to use the SPI port, it needs to be setup right or the SPI hardware doesn't work correctly.  I think this should stay with the configuration code.

aykevl replied:
Yes, but it should use the fixed pins for this chip:

```suggestion
	SPI0_SCK_PIN.Configure(PinConfig{Mode: PinOutput})
	SPI0_SDI_PIN.Configure(PinConfig{Mode: PinInput})
	SPI0_SDO_PIN.Configure(PinConfig{Mode: PinOutput})
```

aykevl commented on src/machine/machine_avr.go:
```diff
@@ -154,3 +154,6 @@ var (
 	UART0 = UART{Buffer: NewRingBuffer()}
 )
+
+// SPI on AVR.
+var SPI0 = SPI{}
```
This should be in the same file as where the SPI is defined, as SPI on AVR depends on fixed pin numbers.

aykevl commented on src/examples/spi/spi.go:
```diff
@@ -0,0 +1,29 @@
+	"time"
+)
+
+func main() {
+	cs := machine.SPI0_CS_PIN
+	cs.High()
+	cs.Configure(machine.PinConfig{Mode: machine.PinOutput})
```
@deadprogram do you have an opinion on whether to include an example here?

In any case, if you include an example you should also add a line to the smoke tests in the Makefile to make sure it continues to compile.

schmidtw replied:
Happy to add to the smoke tests.  I apologize for the broken build & missing this.  I tried running the tests, but since I didn't rebuild llvm from scratch something in the make file doesn't seem happy.  I'm working on running them one by one.

aykevl replied:
You don't need to build with a custom LLVM build. At least if you do `make smoketest TINYGO=tinygo` you can pick the binary that you'd like to use.

deadprogram replied:
Seems like there will be no need for a specific example, by standardizing a bit more with the other SPI implementations. Specifically, removing  `[]CSPins`.

aykevl reviewed (commented):
I think the SPI implementation should live in machine_atmega328p.go, and definitions like `SPI0_SCK_PIN` too. The reason is that the ports used for SPI cannot be configured in software, unlike many other (more powerful) microcontrollers. That should fix errors like the following:

```
/go/bin/tinygo build -size short -o test.hex -target=atmega1284p         examples/serial
# machine
/home/circleci/project/src/machine/machine_atmega.go:181:16: undeclared name: SPI0_SCK_PIN
/home/circleci/project/src/machine/machine_atmega.go:184:16: undeclared name: SPI0_SDO_PIN
/home/circleci/project/src/machine/machine_atmega.go:187:16: undeclared name: SPI0_SDI_PIN
```

aykevl commented:
> And just to be sure - have you made sure that your "tinygo flash ..." commands refers to the correct tinygo version, that you have built?
> What i do most of the time to make sure it uses the correct tinygo binary, i build tinygo with the -o flag like
> 
> ```shell
>  go build -o my_tinygo
> ```

You don't need to build TinyGo each time you make a change to the machine package, only when you make a change to the compiler.  The standard library is not included in the `tinygo` binary.

schmidtw commented on src/machine/machine_atmega.go:
```diff
@@ -155,3 +157,109 @@ func (uart UART) WriteByte(c byte) error {
+// Configure is intended to setup the SPI interface.
+func (s SPI) Configure(config SPIConfig) error {
+
+	// This is only here to help catch a bug with the configuration
```
I ran into missing configuration the first time I added a chip.  It was pretty annoying, so I decided I'd try to detect and return an error.  If you don't like this I can delete it with no hurt feelings.

deadprogram replied:
Having `Configure()` return an error is a standardization we are doing for all I2C interfaces, so seems like a good thing here as well.

deadprogram commented on src/machine/machine_atmega.go:
```diff
@@ -155,3 +157,109 @@ func (uart UART) WriteByte(c byte) error {
+
+	// Ensure the CS pins are all high before enabling them to prevent any
+	// momentary low glitches that can confuse the other chips
+	for _, cs := range config.CSPins {
```
I do not really see what benefit having the `CSPins` slice provides here. Especially since it is not used anywhere else.

schmidtw replied:
It's not included elsewhere but some place needs to be the first :).  (There isn't a governing interface for SPI that I could see).

I'll argue that the CS pins are missing from the other impls and not the other way around.  Here are a few examples of code that all have a bug on Atmega chips that may or may not have issues on different chips:

- https://github.com/tinygo-org/drivers/blob/e376785596dc8269f3e8aa42a9bf75fb1457febc/wifinina/wifinina.go#L271
- https://github.com/tinygo-org/drivers/blob/e376785596dc8269f3e8aa42a9bf75fb1457febc/flash/transport_spi.go#L44
- https://github.com/tinygo-org/drivers/blob/e376785596dc8269f3e8aa42a9bf75fb1457febc/bmi160/bmi160.go#L31

The output pin goes low for a few ns & then high.  For some chips this can signal the start of a transaction & inadvertently cause the first few attempts to use a chip not work.

I definitely agree with standardizing the SPI port usage.  I'm hopeful that driver or application code can delegate pin configuration to the SPI code for the chip (like it does for the other SPI pins).

But if this is the only thing keeping this code from being merged I'll remove it.

schmidtw replied:
@deadprogram If the only thing blocking this merge is the CS handling - I'll remove it to get the rest of the code in.  I'm fine dealing with it in a later PR as well.  I'm only pinging you here because there's so many comments & I don't want you to have to waste your time looking through them if this is the only blocker.

deadprogram replied:
I think we should probably not try to make changes to the interface for all SPI as part of this PR.

That would better served as part of a broader conversation of possible changes to the SPI interface.

deadprogram commented on src/machine/machine_atmega.go:
```diff
@@ -155,3 +157,109 @@ func (uart UART) WriteByte(c byte) error {
+	Frequency uint32
+	LSBFirst  bool
+	Mode      uint8
+	CSPins    []Pin
```
This should probably be removed. See my comment below.

deadprogram commented:
Just tried to test this branch using my Arduino Uno and a small strip of APA102 LED's with the example at https://github.com/tinygo-org/drivers/blob/release/examples/apa102/main.go

I changed the number of LEDs to 6, as that is all I have in my tiny strip. At first it did not work at all. When I removed the params from the `SPIConfig{}` to use the defaults, if I do a hard power reset it will light up the strip once sometimes. The times when it lights up, the animated effect does not occur.

schmidtw commented:
I'll look into this.

schmidtw commented:
I looked at the output of the SPI pins with my logic analyser & based on what I see the code wanting to send out the SPI output matches.  I'm not an APA102 export nor do I have any of those.

This is with my Arduino Uno.  I modified the program to only represent 2 LEDs:
```
	leds := make([]color.RGBA, 2)
```

Here is the output waveform.  The clock is 500kHz.  I added a CS so I could easily trigger the analyser.

![Screenshot from 2021-02-09 11-49-44](https://user-images.githubusercontent.com/698825/107419990-45fd4c00-6acd-11eb-9015-efaaab42ae30.png)

As I wrote this up I decided to re-run and capture without the CS pin being configured.  The SPI output is not there.  So what I have found is that the CS pin configuration is critical for the SPI peripheral hardware to actually work.  The Atmel datasheets indicate that if the specified (default) CS pin is not in the right state, the SPI peripheral will start in slave mode, not master mode.

I added this to the example right before the `apa102.New()` is called & the waveform works as expected:
```
	cs := machine.PB2
	cs.High()
	cs.Configure(machine.PinConfig{Mode: machine.PinOutput})
```

Here's what it looks like with the CS pin configured, but not used (the SPI decoder didn't run since I have it configured to require the CS signal but the waveform is the same):
![Screenshot from 2021-02-09 12-00-14](https://user-images.githubusercontent.com/698825/107421059-6548a900-6ace-11eb-8729-2a6f8baf004c.png)

deadprogram commented:
That is pretty interesting. Confirmed on my end by adding the cs pin as you suggested,

Hmm, AVR SPI is pretty specific about its requirements it seems.

schmidtw commented:
Since the application actually controls the CS operation for AVRs the important thing is that during configuration/enabling of the peripheral the CS is in the right state.  After that, it does not matter what is used for CS if one is used at all.  In theory you could use the pin for something else, but that will be quite complex to manage & shouldn't live in this code.

I tried adding the least invasive change possible in the latest change I pushed.

deadprogram commented:
Something else: the APA102 freezes after about 10-15 seconds.

Keep in mind that this APA102 is one of my main devices I use for SPI testing on every other platform, using that same code in the drivers repo. So hopefully the same will be able to be true for AVR.

deadprogram commented:
That last commit you pushed seems to have resolved that freeze. That CS pin really means it! :smile_cat:

deadprogram commented:
Was also able to run the MCP3008 example on Arduino Uno: https://github.com/tinygo-org/tinygo/blob/release/src/examples/mcp3008/mcp3008.go :tada:

deadprogram commented:
Also confirmed MCP3008 example on Arduino Mega 2560.

@schmidtw if you do not mind, could you please squash the commits in this PR so I can merge it now? 

Thank you so much for working on this, SPI on AVR has been a long needed feature!

deadprogram commented:
Let me also not forget @Nerzal who did so much work on this to get it started! A true group effort.

schmidtw committed:
Add SPI support for Atmega based chips.

This is based on @Nerzal's #1398 PR, but is a bit of a refactor and
expansion to support all the Atmega based chips present in tinygo.

Nerzal commented:
Should we add the info, that the CS pin needs to be configured prior to configure the SPI interface to the tinygo-site ? 
Or is this "common sense" ?

deadprogram commented:
Now merging, thanks again everyone!

aykevl commented:
Sorry to be so late in the review. I think the code can be further simplified so that it takes a lot less space (at the cost of being less portable). Maybe I can do that in a new PR? The current solution with a global with pointers will be rather space inefficient, something that is a problem on microcontrollers.

deadprogram commented:
That sounds very good to me!
